### PR TITLE
Java: CWE-798 Query to detect hard-coded Azure credentials

### PR DESCRIPTION
**Pull request by luchua-bc**
Microsoft Azure is one of the most popular cloud computing solutions for building, testing, deploying, and managing applications and services in the cloud.

Azure offers a well-maintained [Java SDK](https://github.com/Azure/azure-sdk-for-java) for provisioning, managing, and using Azure resources from Java application code. The Azure SDK for Java is composed of many individual Java libraries that relate to specific Azure services.

The query detects calling Azure SDK with a hard-coded user name and password or client secret. 

Please consider to merge the PR. Thanks.

### CODE REVIEW

File: java/ql/src/Security/CWE/CWE-798/HardcodedCredentialsApiCall.qhelp [hunk 1]
```diff
@@ -32,13 +32,43 @@
...
+  <p>
+    The following code example connects to Azure using a hard-coded user name and password or client secret:
+  </p>
+
+  <sample src="HardcodedAzureCredentials.java"/>
+
+  <p>
+    Instead, the username and password or client secret could be supplied through environment variables,
+    which can be set externally without hard-coding credentials in the source code.
+  </p>
```
Comment (smowton) [inline java/ql/src/Security/CWE/CWE-798/HardcodedCredentialsApiCall.qhelp:56]: No need to change the qhelp, the query is already generically trying to find hardcoded creds; there's no need for it to exhaustively show each one.

**Review by smowton — COMMENTED** [review]
Add a change note

**Comment by luchua-bc** [comment]
Thanks @smowton for reviewing this PR. I've made all requested changes. Please re-review.

File: java/ql/src/Security/CWE/CWE-798/HardcodedCredentialsApiCall.qhelp (hunk 1 above)
Comment (luchua-bc) [reply java/ql/src/Security/CWE/CWE-798/HardcodedCredentialsApiCall.qhelp]: I've added a change note and reset the qhelp file.

File: java/ql/src/Security/CWE/CWE-798/HardcodedAzureCredentials.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+public class HardcodedAzureCredentials {
```
Comment (smowton) [inline java/ql/src/Security/CWE/CWE-798/HardcodedAzureCredentials.java:1]: File no longer used

**Comment by smowton** [comment]
@aschackmull I note two of the predicates in `SensitiveApi.qll` say `// Auto-generated using an auxiliary query run on the JDK source code.`. Is it acceptable to add to the end of that list or is that aux query re-run periodically with the risk of overwriting these changes? Perhaps we should define additional predicates for manually-added methods outside the JDK?

File: java/ql/src/Security/CWE/CWE-798/SensitiveApi.qll [hunk 3]
```diff
@@ -202,7 +203,8 @@ private predicate javaApiCallableUsernameParam(string s) {
   s = "sun.tools.jconsole.ProxyClient;getConnectionName(String, String);1" or
   s = "sun.tools.jconsole.ProxyClient;getProxyClient(String, int, String, String);2" or
   s = "sun.tools.jconsole.ProxyClient;getConnectionName(String, int, String);2" or
-  s = "com.amazonaws.auth.BasicAWSCredentials;BasicAWSCredentials(String, String);0"
+  s = "com.amazonaws.auth.BasicAWSCredentials;BasicAWSCredentials(String, String);0" or
```
Comment (pwntester) [inline java/ql/src/Security/CWE/CWE-798/SensitiveApi.qll:206]: The comment above suggests this list is Auto-generated. @smowton will this new addition get overriden if placed here?

File: java/ql/src/Security/CWE/CWE-798/HardcodedAzureCredentials.java (hunk 2 above)
Comment (luchua-bc) [reply java/ql/src/Security/CWE/CWE-798/HardcodedAzureCredentials.java]: Removed the sample Java file together with the <code>.qhelp</code> help file

File: java/ql/src/Security/CWE/CWE-798/SensitiveApi.qll (hunk 3 above)
Comment (luchua-bc) [reply java/ql/src/Security/CWE/CWE-798/SensitiveApi.qll:206]: @smowton  and @pwntester - shall we move these APIs of Azure together with those of AWS to the end of
https://github.com/github/codeql/blob/4d014717b6a109d61ae3f9c803092f8df27ede07/java/ql/src/Security/CWE/CWE-798/SensitiveApi.qll#L509-L516?

Comment (smowton) [reply java/ql/src/Security/CWE/CWE-798/SensitiveApi.qll:206]: Yes that looks like a good idea, don't modify any of the predicates flagged autogen'd. They all get OR'd together anyhow:

```
        javaApiCallableUsernameParam(call.getCallee(), i) or
        javaApiCallablePasswordParam(call.getCallee(), i) or
        javaApiCallableCryptoKeyParam(call.getCallee(), i) or
        otherApiCallableCredentialParam(call.getCallee(), i)
```

Comment (luchua-bc) [reply java/ql/src/Security/CWE/CWE-798/SensitiveApi.qll]: Thanks @smowton for the confirmation. I've made the change.

File: java/change-notes/2021-05-12-hardcoded-azure-credentials-in-api-call.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+lgtm,codescanning
+* The query "Hard-coded credential in API call" (`java/hardcoded-credential-api-call`)
+  now recognizes hard-coded authentication credentials with Azure SDK for Java.
```
Comment (smowton) [inline java/change-notes/2021-05-12-hardcoded-azure-credentials-in-api-call.md:3]:
```suggestion
  now recognizes hard-coded authentication credentials passed to the Azure SDK for Java.
```

**Review by smowton — COMMENTED** [review]
lgtm; over to @aschackmull for final review

File: java/change-notes/2021-05-12-hardcoded-azure-credentials-in-api-call.md (hunk 4 above)
Comment (luchua-bc) [reply java/change-notes/2021-05-12-hardcoded-azure-credentials-in-api-call.md]: Updated

**Comment by luchua-bc** [comment]
@aschackmull @smowton I'll close this PR once @bananabr or another staff merges this query. Or please close it on my behalf.

**Review by smowton — APPROVED** [review]

**Comment by smowton** [comment]
Sorry, forgot about this because it wasn't in the bug bounty program any longer

**Comment by luchua-bc** [comment]
@smowton Thanks for merging the code and closing the PR. I know it wasn't in the bug bounty program any longer as per the previous discussion and just want to have the code contributed to the code repository to have one less open PRs:-)

**Comment by bananabr** [comment]
Hi @luchua-bc,

I can definitely work on merging this query. I had an impossible week but I plan to work on my PR soon.

**Comment by luchua-bc** [comment]
Thanks @bananabr for the update.

**Pull request merged**